lms/services/course_roster.py

```diff
@@ -1,49 +1,140 @@
+from lms.services.lti_role_service import LTIRoleService
+from lms.services.upsert import bulk_upsert
 
 LOG = getLogger(__name__)
 
 
 class CourseRosterService:
-    def __init__(self, db, lti_names_roles_service: LTINamesRolesService):
+    def __init__(
+        self,
+        db,
+        lti_names_roles_service: LTINamesRolesService,
+        lti_role_service: LTIRoleService,
+        h_authority: str,
+    ):
         self._db = db
         self._lti_names_roles_service = lti_names_roles_service
+        self._lti_role_service = lti_role_service
+        self._h_authority = h_authority
 
     def fetch_roster(self, lms_course: LMSCourse) -> None:
+        assert (
+            lms_course.lti_context_memberships_url
+        ), "Trying fetch roster for course without service URL."
+
         lti_registration = self._db.scalars(
             select(LTIRegistration)
             .join(ApplicationInstance)
-            .join(LMSCourseApplicationInstance)
             .where(LMSCourseApplicationInstance.lms_course_id == lms_course.id)
+            .join(LMSCourseApplicationInstance)
             .order_by(LTIRegistration.updated.desc())
         ).first()
-
         assert lti_registration, "No LTI registration found for LMSCourse."
-        assert (
-            lms_course.lti_context_memberships_url
-        ), "Trying fetch roster for course without service URL."
 
         roster = self._lti_names_roles_service.get_context_memberships(
             lti_registration, lms_course.lti_context_memberships_url
         )
-        LOG.info(
-            "Roster for %s. Users returned %s",
-            lms_course.h_authority_provided_id,
-            len(roster),
+
+        # Insert any users we might be missing in the DB
+        lms_users_by_lti_user_id = {
+            u.lti_user_id: u
+            for u in self._get_roster_users(
+                roster, lms_course.tool_consumer_instance_guid
+            )
+        }
+        # Also insert any roles we might be missing
+        lti_roles_by_value: dict[str, LTIRole] = {
+            r.value: r for r in self._get_roster_roles(roster)
+        }
+
+        # Make sure any new rows have IDs
+        self._db.flush()
+
+        roster_upsert_elements = []
+
+        for member in roster:
+            lti_user_id = member.get("lti11_legacy_user_id") or member["user_id"]
+            # Now, for every user + role, insert a row  in the roster table
+            for role in member["roles"]:
+                roster_upsert_elements.append(
+                    {
+                        "lms_course_id": lms_course.id,
+                        "lms_user_id": lms_users_by_lti_user_id[lti_user_id].id,
+                        "lti_role_id": lti_roles_by_value[role].id,
+                        "active": member["status"] == "Active",
+                    }
+                )
+        # We'll first mark everyone as non-Active. We keep a record of how belonged to a course.
```

Comment:
I guess you meant `who`, not `how`.

```suggestion
        # We'll first mark everyone as non-Active. We keep a record of who belonged to a course.
```